rust/rope/src/interval.rs

```diff
@@ -12,109 +12,83 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//! Intervals that can be open or closed at the ends.
+//! Closed-open intervals, and operations on them.
+
+
+//NOTE: intervals used to be more fancy, and could be open or closed on either
+//end. It may now be worth considering replacing intervals with Range<usize> or similar.
 
 use std::cmp::{min, max};
 use std::fmt;
+use std::ops::{Range, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive};
 
-// Invariant: end >= start
-// (attempting to construct an interval of negative size gives an
-// empty interval beginning at start)
-
+/// A fancy version of Range<usize>, representing a closed-open range;
+/// the interval [5, 7) is the set {5, 6}.
+///
+/// It is an invariant that `start <= end`. An interval where `end < start` is
+/// considered empty.
 #[derive(Clone, Copy, PartialEq, Eq)]
 pub struct Interval {
-    start: u64,  // 2 * the actual value + 1 if open
-    end: u64,    // 2 * the actual value + 1 if closed
-}
-
-impl fmt::Display for Interval {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        if self.is_start_closed() {
-            write!(f, "[")?;
-        } else {
-            write!(f, "(")?;
-        }
-        write!(f, "{}, {}", self.start(), self.end())?;
-        if self.is_end_closed() {
-            write!(f, "]")?;
-        } else {
-            write!(f, ")")?;
-        }
-        Ok(())
-    }
-}
-
-impl fmt::Debug for Interval {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Display::fmt(self, f)
-    }
+    pub start: usize,
+    pub end: usize,
 }
 
 impl Interval {
-    pub fn new(start: usize, start_closed : bool, end: usize, end_closed: bool) -> Interval {
-        let start = (start as u64) * 2 + if start_closed { 0 } else { 1 };
-        let end = (end as u64) * 2 + if end_closed { 1 } else { 0 };
+    pub fn new(start: usize, end: usize) -> Interval {
```

Comment:
This is the place to decide if there's an end >= start invariant. My feeling is yes; so it should be added as a docstring and a debug assert.